ProblyWorks: Conformal Prediction (implementation APS and LAC)

ilknur12 opened the pull request:
## Issue

Pull Request for Conformal Prediction APS and LAC implementation in probly:

**Involved issues:**

_Build API_: https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142335388&issue=Cortys%7CSEP-Probly-WS2526%7C322
_Test API:_ https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142339337&issue=Cortys%7CSEP-Probly-WS2526%7C324

**APS:** 
_notebook(with probly):_ https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=143790173&issue=Cortys%7CSEP-Probly-WS2526%7C373

`common.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142343035&issue=Cortys%7CSEP-Probly-WS2526%7C325
`test_common.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142351049&issue=Cortys%7CSEP-Probly-WS2526%7C331

`torch.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142343575&issue=Cortys%7CSEP-Probly-WS2526%7C326
`test_torch.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142351246&issue=Cortys%7CSEP-Probly-WS2526%7C332

`flax.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142344085&issue=Cortys%7CSEP-Probly-WS2526%7C327
`test_flax.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142351706&issue=Cortys%7CSEP-Probly-WS2526%7C333

**APS method:**
`split_conformal.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142352697&issue=Cortys%7CSEP-Probly-WS2526%7C335
`test_split_conformal.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142362225&issue=Cortys%7CSEP-Probly-WS2526%7C346

**LAC:**
_notebook (with probly):_ https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=143790153&issue=Cortys%7CSEP-Probly-WS2526%7C372

`common.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142356531&issue=Cortys%7CSEP-Probly-WS2526%7C336
`test_common.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142360348&issue=Cortys%7CSEP-Probly-WS2526%7C343

`torch.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142357289&issue=Cortys%7CSEP-Probly-WS2526%7C337
`test_torch.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142360805&issue=Cortys%7CSEP-Probly-WS2526%7C344

`flax.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142357456&issue=Cortys%7CSEP-Probly-WS2526%7C338
`test_flax.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142360992&issue=Cortys%7CSEP-Probly-WS2526%7C345

**LAC method:** 
`accretive_completion.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142358294&issue=Cortys%7CSEP-Probly-WS2526%7C339
`test_accretive_completion.py:` https://github.com/users/Cortys/projects/4/views/7?sliceBy%5Bvalue%5D=8%3A+ProblyWorks&pane=issue&itemId=142360100&issue=Cortys%7CSEP-Probly-WS2526%7C342


## Motivation and Context

With this PR we implemented APS and LAC for our Conformal Prediction for probly. We tried to have a clean structure and flexible code.

## Public API Changes

-   [X] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

The files were tested locally with pre-commit and pytest.
We also did a test-PR run in our own fork in which all tests have succeeded: https://github.com/m3ryemmo/probly/pull/18

---

## Checklist

-   [X] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [X] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [X] I have considered the impact of these changes on the public API.

---

alireza-javanmardi reviewed (commented):
cp/aps/methpds/common.py

alireza-javanmardi commented on src/probly/conformal_prediction/aps/common.py:
```diff
@@ -0,0 +1,61 @@
+        true_label_pos = sorted_indices.index(labels[i])
+        scores[i] = cumulative_probs[true_label_pos].item()
+
+    return scores
```
This is not exactly the APS score; rather, it is just the cumulative probability score. APS includes a random term (“u” in the paper), which is important to incorporate. 
Specifically, something like **scores -= u * probs** is needed, where u is a vector of independent uniform random variables.

Though I would make this "u" term optional.

alireza-javanmardi commented on src/probly/conformal_prediction/aps/torch.py:
```diff
@@ -0,0 +1,155 @@
+        self,
+        x: Sequence[Any],
+        significance: float,
+    ) -> list[list[int]]:
```
Instead of outputting a list, please output an array of size n_samples × n_classes with binary values (0: not in the set, 1: in the set).

alireza-javanmardi commented on src/probly/conformal_prediction/aps/torch.py:
```diff
@@ -0,0 +1,155 @@
+    from collections.abc import Sequence
+
+
+class APSPredictor(ConformalPredictor):
```
We do not need many nonconformity classes. 

Basically, for any nonconformity score, we require access to the model probabilities, then we calculate the nonconformity scores for all calibration and test points, compute the quantile, and finally provide the prediction sets. Therefore, we only need a single class with pluggable nonconformity score functions.

ilknur12 commented:
Hey! We refactored our code and implemented the requested changes.

**Changes:**
- created a generic nonconformity score class in already existing `scores/common.py,` which handles the label extraction, prediction logic and randomization
    -> reduced code redundancy in the score implementations
- added new nonconformity scores for classification (`SAPS`, `RAPS`) and for regression (`CQR`, `Absolute Error`) with the same structure as the other scores before
- corrected the `for` loop in `metrics`
- added new predictors: `Mondrian` and its special case `Class Conditional` as well as `Jackknife+` and `CV+`

We merged the changes via a PR made in our own fork: https://github.com/m3ryemmo/probly/pull/22.

Cortys commented on src/probly/conformal_prediction/notebooks/aps.ipynb:
move this to `notebooks/examples` in root (should not be in `src`)
